dvc/output.py

```diff
@@ -330,11 +338,26 @@ def __init__(
         self.live = live
         self.desc = desc
 
-        self.fs_path = self._parse_path(self.fs, fs_path)
+        self._fs_path = self._parse_path(self.fs, fs_path)
         self.obj = None
 
         self.remote = remote
 
+        if self.fs.version_aware:
+            self.def_path, version_id = self.fs.path.coalesce_version(
+                self.def_path, self.meta.version_id
+            )
+            self._fs_path, _ = self.fs.path.split_version(self._fs_path)
+            self.meta.version_id = version_id
+
+    @cached_property
+    def fs_path(self) -> str:
+        if self.fs.version_aware and self.meta.version_id:
+            return self.fs.path.version_path(
+                self._fs_path, self.meta.version_id
+            )
+        return self._fs_path
```

Comment:
If we have `meta.version_id`:
- `output.fs_path` will be versioned if `output.fs` supports versioning
- If the fs does not support versioning (like when we have a local output with saved `version_id` from some cloud), `output.fs_path` will not be versioned

`output.def_path` and `output._fs_path` are always unversioned.

This makes all existing fs operations work as expected (i.e. `output.download()` will download using the versioned URL).

When we need an explicitly unversioned path (like when checking to see if an import update is available), we can use `output._fs_path` directly.

Reply:
This does not handle the new workflow situation where we have local output with a cloud `version_id`, and need to tie that version ID with the source path during fetch (handling this probably goes in dvc-data)

Reply:
It seems to me that `path` handling should not even be part of the `Output`. Also output is very loaded.

Reply:
How does this work with directories btw?

Reply:
Also, maybe this versioning thing is a different kind of `Output`? Have you thought in that way?

Reply:
We only support versioning for files right now, directories will always have a `None` version ID

Reply:
@pmrowla Hm, didn't we agree yesterday that fs_path should not have version id in it? That will really mess with index collection.